Add new DV reason: ImagePullFailed

nunnatsa opened the pull request:
## What this PR does / why we need it
When the image pull fails, the DataVolume Running condition, will have the Reason field of `ImagePullFailed`, to allow better error handling in code that uses it.

Also, moved the `StreamDataToFile` function from the `util` package to the `importer` package, and made to non-exported. This is done to allow calling the new `NewErrImagePullFailed` function from within `streamDataToFile`, and to avoid import cycle. In addition, the `streamDataToFile` is only used in the `importer` package and so it's better to define it there.

### Release note
```release-note
When the image pull fails, the DataVolume Running condition, will have the Reason field of "ImagePullFailed"
```

```jira-ticket
https://issues.redhat.com/browse/CNV-39603
```

nunnatsa commented:
/test pull-containerized-data-importer-e2e-ceph-wffc

awels commented on cmd/cdi-importer/importer.go:
```diff
@@ -179,6 +180,13 @@ func handleImport(
+
+	ipfErr := &importer.ErrImagePullFailed{}
+	isImagePullFailure := errors.As(err, &ipfErr)
+	if isImagePullFailure {
```
Why not just do
```go
if errors.As(err, &ipfErr) {
  err = ipfErr
}
```
Also if the `As` is true, isn't err already the right type? I am not entirely I understand what is going on here.

nunnatsa replied:
We don't need the boolean here. leftover from another implementation during the development. Will fix.

The `As` function does not change `err` but only populate `ipfErr`

nunnatsa replied:
In 2nd thought, I don't need this code at all. I was trying to write a new field in the `TerminationMessage` object, but this is not working anyway for other reasons. 

I'm now relaying on the error text, that is still there.

nunnatsa replied:
removed this code

awels commented on tests/import_test.go:
```diff
@@ -89,6 +90,34 @@ var _ = Describe("[rfe_id:1115][crit:high][vendor:cnv-qe@redhat.com][level:compo
+
+	},
+		Entry("insecure registry", "registry:5000/myimage:latest"),
+		Entry("non exists image", "quay.io/capk/ubuntu-2004-container-disk:wrong"),
```
Should not be a real issue, but maybe reference the internal registry here with an invalid image. This now assumes you have access to the internet, which may not always be valid.

nunnatsa replied:
Any other idea?

awels reviewed (commented):
Looks like most of this PR is not actually related to returning the error message, but in moving the StreamDataToFile function. Other than that some minor comments.

akalenyu reviewed (commented):
Thanks!
Had some minor comments but what's more interesting to me is why we can't use the termination message API in case of failures.
Is there something missing in it's implementation? I won't ofcourse block over this

nunnatsa commented:
> Thanks! Had some minor comments but what's more interesting to me is why we can't use the termination message API in case of failures. Is there something missing in it's implementation? I won't ofcourse block over this

the termination message API (object) is only available for pod that was terminated with no error. For error cases, the pod write unstructed text, and I didn't want to change this behavior because it's way out of the scope of this PR.

nunnatsa commented on tests/import_test.go:
```diff
@@ -89,6 +90,34 @@ var _ = Describe("[rfe_id:1115][crit:high][vendor:cnv-qe@redhat.com][level:compo
+		utils.WaitForConditions(f, dv.Name, f.Namespace.Name, controllerSkipPVCCompleteTimeout, assertionPollInterval, runningCondition)
+
+	},
+		Entry("insecure registry", "registry:5000/myimage:latest"),
```
I wanted to catch to cases here that should be easy to reproduce: insecure registry and non exist image tag. But I don't think the test will fail outside of kvci, because it will still won't be able to pull the image, for other reasons (e.g. registry not found)

akalenyu replied:
yeah it fails on
```
transport.go:78] Could not create image reference: pinging container registry registry:5000: Get "https://registry:5000/v2/": dial tcp: lookup registry on 172.30.0.10:53: no such host
```
so should return your new err.

but how come this tests insecure registry? AFAIK the kubevirtci VM has this registry whitelisted

nunnatsa replied:
There is a field in CDI CR for that. I'm not sure what is it setting in kvci, but locally this is what I'm getting.

nunnatsa commented:
> e, but in moving the StreamDataToFile function. Other than that some minor comments.

Eventually, yes. But this is not the "why". I moved the function because I wanted to call the `NewErrImagePullFailed` function from it, and this function is defined in the importer package. Then I found out that the `StreamDataToFile` function only called from the importer package. So to avoid import cycle, I moved the function, but this is not the reason for this change.

nunnatsa commented on tests/import_test.go:
```diff
@@ -89,6 +90,34 @@ var _ = Describe("[rfe_id:1115][crit:high][vendor:cnv-qe@redhat.com][level:compo
 		Entry("fail given a large physical size QCOW2 file", utils.LargePhysicalDiskQcow),
 	)
 
+	DescribeTable(`Should fail with "ImagePullFailed" reason if failed to pull image`, func(url string) {
```
done

nunnatsa commented:
/retest

akalenyu reviewed (commented):
Looks good! thanks

Correct me if I'm wrong but this PR does not handle `pullMethod: node` (this is fine),
which I thought is quite common on OpenShift. Are you okay with that?

nunnatsa commented:
^^^ fixed new linter warning

nunnatsa commented:
/retest

akalenyu commented on tests/import_test.go:
```diff
@@ -2199,6 +2200,41 @@ var _ = Describe("Containerdisk envs to PVC labels", func() {
+
+	},
+		Entry("pull method = pod", "myregistry/myorg/myimage:wrongtag", cdiv1.RegistryPullPod),
+		Entry("pull method = node", "myregistry/myorg/myimage:wrongtag", cdiv1.RegistryPullNode),
```
:+1:

akalenyu reviewed (commented):
/approve

nunnatsa commented:
/test pull-containerized-data-importer-e2e-upg

awels commented:
/lgtm
/approve
/cherrypick release-v1.59

nunnatsa commented:
/test pull-containerized-data-importer-e2e-upg
